chore(Node): Do not keep the entire genesis file in memory during node operation

alesforz opened the pull request:
Closes #1290.

This PR tracks the work we are doing for #1290. 
For details, please refer to #1290 discussion.

### ToDo
- [x] Environment stores either a pointer to a GenesisDoc or the genesis' chunks, but not both. Merged to this branch with #4235.
- [x] [Node](https://github.com/cometbft/cometbft/blob/87a8c8948bf9926aff37f344b4120163f12cb4ad/node/node.go#L50) does not keep a pointer to a GenesisDoc after initialization. Merged to this branch with #4250.

---

#### PR checklist

- [x] Tests written/updated
- [x] Changelog entry added in `.changelog` (we use [unclog](https://github.com/informalsystems/unclog) to manage our changelog)
- [x] Updated relevant documentation (`docs/` or `spec/`) and code comments

alesforz commented:
See [here](https://github.com/cometbft/cometbft/issues/1290#issuecomment-2417082937) for a visualization of the changes in memory usage after applying these changes.

alesforz commented:
Note: despite our efforts, these changes break the `Node` public API. Because `Node` doesn't store the genesis in memory after startup, its [exported method](https://github.com/cometbft/cometbft/blob/4eb7c058dd99c092b932c79d7cae33010f57cad9/node/node.go#L989) `GenesisDoc()` had to [change](https://github.com/cometbft/cometbft/blob/6bfe1e15e06ca3581fdb392adddb91e9a1ca7b68/node/node.go#L1038):
- it now fetches the genesis from disk and unmarshals it into a `*types.GenesisDoc`. 
- its signature changed to also return an error in case fetching the genesis from disk fails (this is the breaking change)

jmalicevic commented on rpc/client/local/local.go:
```diff
@@ -50,6 +50,7 @@ func New(node *nm.Node) *Local {
 	if err != nil {
 		node.Logger.Error("Error configuring RPC", "err", err)
 	}
+
```
```suggestion
```

jmalicevic commented on rpc/core/env.go:
```diff
@@ -100,6 +107,69 @@ type Environment struct {
+	// If genesis is less than 16MB, then no chunking.
+	// Keep a pointer to a GenesisDoc in env.GenDoc.
+	if len(data) <= genesisChunkSize {
+		env.genChunks = nil
```
But the `genesis_chunked` endpoint still remains active and returns the chunks ? So we either edit that to create a on-demand chunk, or we do not delete the chunks here. I think , for now, it is simpler to just not set it to nil. That is, if we have only one chunk (<16MB) , then its not that big of an overhead if we keep it in memory.

alesforz replied:
The `genesis_chunked` endpoint remains active but returns an error.
See [here](https://github.com/cometbft/cometbft/blob/6bfe1e15e06ca3581fdb392adddb91e9a1ca7b68/rpc/core/net.go#L121). I changed the endpoint so that:
- if there are no chunks but the request's chunk id = 0, then it returns the genesis serialized to JSON. This preserves the current behavior of `/genesis_chunked`
- if there are no chunks and the requested chunk id >= 1, it returns an error. This is also keeps the existing behavior.

This way, we can ditch the chunk's slice.

jmalicevic replied:
Ah I missed that for some reason even though I searched for it . That's ok then. I guess you are just delaying the chunk creation until it is actually needed and then it is created only once.

jmalicevic commented on rpc/core/env.go:
```diff
@@ -100,6 +107,69 @@ type Environment struct {
+
+	env.genChunks = chunks
+
+	// we store the chunks; don't store a ptr to the genesis anymore.
+	env.GenDoc = nil
+
```
If you apply my comment above than this has to be changed to 
```suggestion
	// we store the chunks; don't store a ptr to the genesis anymore.
	if len(chunks) > 1{
		env.GenDoc = nil
	}

```

alesforz replied:
see my reply above 😄

andynog reviewed (approved):
Hi @alesforz, I think the implementation looks good, I've added comments and suggestions. I've also tested the API using a sample file and seems to work. 

I didn't not validate the content returned. Just realized that we should provide maybe a CLI command (or sample app) in the future that could download the chunks and re-assemble into a genesis doc. Expecting users and operators to download encoded chunks might not be the best UX

jmalicevic commented on rpc/core/env_test.go:
```diff
@@ -2,12 +2,150 @@ package core
+	// nil, chunks slice stored in genChunks field.
+	// The test genesis file is the osmosis chain genesis (~69MB).
+	t.Run("Chunking", func(t *testing.T) {
+		const fGenesisPath = "./testdata/genesis_osmosis.json"
```
I agree that it might not be the best idea to include the actual genesis files in the repo just for testing purposes.

alesforz replied:
🤔 you are right. I'll remove it and use a genesis file with a randomly generated `app_state` instead.

jmalicevic commented on rpc/core/net.go:
```diff
@@ -105,15 +107,37 @@ func (env *Environment) UnsafeDialPeers(
+		// See discussion in the following PR for why we still serve chunk 0 even
+		// if env.genChunks is nil:
+		// https://github.com/cometbft/cometbft/pull/4235#issuecomment-2389109521
+		if chunk == 0 {
```
To my understanding this is , from the clients perspective, the same behaviour. For both endpoints.

alesforz replied:
It is not new behavior. The current implementation of `InitGenesisChunks()` creates the `genChunks` slice even when the genesis's size is <= `genesisChunkSize` (i.e., chunks aren't needed). In this case, the `genChunks` slice has only 1 element at index 0: the entire genesis encoded as a base64 string. If you query the `/genesis_chunked` endpoint with chunk id = 0, it will return you the entire genesis encoded as base64.

The code you are seeing here preserves this behavior after our changes.

andynog replied:
sounds good, thanks for the clarification

alesforz commented on rpc/core/env.go:
```diff
@@ -100,6 +107,69 @@ type Environment struct {
+// InitGenesisChunks checks whether it makes sense to create a cache of chunked
+// genesis data. It is called on Node startup and should be called only once.
+// Rules of chunking:
+// - if the genesis file's size is <= 16MB, then no chunking. An `Environment`
```
good point 👍

alesforz commented on node/node.go:
```diff
@@ -985,9 +1027,31 @@ func (n *Node) PrivValidator() types.PrivValidator {
+// Note that the genesis file can be large (hundreds of MBs, even GBs); therefore,
+// we recommend that the caller does not keep the GenesisDoc returned by this
+// function in memory longer than necessary.
+func (n *Node) GenesisDoc() (*types.GenesisDoc, error) {
```
Not in our codebase. Tbh, my first instinct was to delete it. However, because this is an exported method of an exported type, there might be users relying on it. Therefore, I left it there, but changed its signature to adapt it to the changes.

It's still a breaking change, but less breaking than removing the method altogether.

andynog replied:
sounds good, thanks for the clarification

alesforz commented on node/node.go:
```diff
@@ -727,40 +738,71 @@ func (n *Node) OnStop() {
+		}
 
-		Logger: n.Logger.With("module", "rpc"),
+		if err := _rpcEnv.InitGenesisChunks(); err != nil {
```
I agree with @jmalicevic. 
Though adding a log message is a good idea, I can add it.

andynog replied:
sounds good to me, good points.

alesforz commented on rpc/core/env.go:
```diff
@@ -100,6 +107,69 @@ type Environment struct {
+	if env.GenDoc == nil {
+		// chunks not computed yet, but no genesis available.
+		// This should not happen.
+		return errors.New("pointer to genesis is nil")
```
Good point. I'll change it.

andynog replied:
I think this is better

alesforz commented on rpc/core/env_test.go:
```diff
@@ -2,12 +2,150 @@ package core
+	"github.com/cometbft/cometbft/types"
 )
 
+func TestInitGenesisChunks(t *testing.T) {
```
🤔 interesting idea. I'll add the draft of a test doing this

andynog replied:
sounds good, thanks !

andynog reviewed (approved):
thanks for addressing the comments and suggestions. Looks good to me.

jmalicevic commented:
Are the genesis json files still needed for testing ?As you added strings, can they be removed from the PR ?

alesforz commented:
> Are the genesis json files still needed for testing ?As you added strings, can they be removed from the PR ?

The files are still there to run the unit tests. They are fictitious though: `genesis_ci.json` is the default genesis we generate to run the e2e test `ci.toml`; `genesis_big.json` is  a copy of `genesis_ci.json` to which I added meaningless strings to the `app_state` field in order to test with a big genesis. Neither have real addresses.

I can substitute the small genesis file with a string in the code, but the big genesis file needs to be there to unit test the chunking mechanism. I can reduce its size (it's about 42MB) if that's your concern.

ValarDragon commented:
Have not reviewed the PR, but am looking forward to this feature!

melekes commented on .changelog/unreleased/breaking-changes/4250-node-no-genesis.md:
```diff
@@ -0,0 +1,2 @@
+- `[Node]` no longer stores a pointer to a `types.GenesisDoc` after completing its
```
```suggestion
- `[node]` Don't store a pointer to a `types.GenesisDoc` after the node is running
```

alesforz commented:
> > Are the genesis json files still needed for testing ?As you added strings, can they be removed from the PR ?
> 
> @jmalicevic The files are still there to run the unit tests. They are fictitious though: `genesis_ci.json` is the default genesis we generate to run the e2e test `ci.toml`; `genesis_big.json` is a copy of `genesis_ci.json` to which I added meaningless strings to the `app_state` field in order to test with a big genesis. Neither have real addresses.
> 
> I can substitute the small genesis file with a string in the code, but the big genesis file needs to be there to unit test the chunking mechanism. I can reduce its size (it's about 42MB) if that's your concern.

Updated the testing code to generate the big genesis docs at runtime rather than having a file in the repo 😄

alesforz commented on node/node.go:
```diff
@@ -727,46 +738,78 @@ func (n *Node) OnStop() {
-	}
-	return &rpcCoreEnv, nil
+		if err := _rpcEnv.InitGenesisChunks(); err != nil {
+			errToReturn = fmt.Errorf("setting up RPC API environment: %s", err)
```
I usually avoid starting error messages with phrases like `failed to do...` or `error...` because it's implicit that an error indicates something went wrong. I think  such phrases are redundant. Another consideration is in Go you chain error strings from nested calls. I find that using `error...` repeatedly can result in awkward and verbose messages. For example, in a call stack of three nested functions, you might get:

```
failed to do X: can't do Y: error doing Z: failed to marshal to JSON
```

Since the innermost error already specifies what actually happened (e.g., failing to marshal to JSON), I believe that focusing on what we were doing when the error occurred makes the message clearer. For example:

```
X: Y: Z: failed to marshal to JSON
```

Here, `X`, `Y`, and `Z` represent the actions being performed at each level, not the function names.

What do you think? Do we have an error convention about this?

melekes replied:
Hmm... I never thought about it in that way. Makes sense!

jmalicevic reviewed (approved):
Thanks for this @alesforz  !
